### PR TITLE
Add Beefy & Near client implementations. ICS 02 refactor [CF work]

### PR DESCRIPTION
**Pull request by Wizdave97**
Closes: #XXX

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules). 
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by adizere** [comment]
We'll merge #2280 first, so this PR might need very minimal adjustments after that PR lands in master. In the meantime, is it possible to look into the failing CI @Wizdave97? We'll look at failed CI jobs as well and see what's up!

File: modules/Cargo.toml [hunk 1]
```diff
@@ -68,8 +96,16 @@ tracing-subscriber = { version = "0.3.11", features = ["fmt", "env-filter", "jso
 test-log = { version = "0.2.10", features = ["trace"] }
 modelator = "0.4.2"
 sha2 = { version = "0.10.2" }
-tendermint-rpc = { version = "=0.23.7", features = ["http-client", "websocket-client"] }
-tendermint-testgen = { version = "=0.23.7" } # Needed for generating (synthetic) light blocks.
+tendermint-rpc = { git = "https://github.com/composableFi/tendermint-rs", branch = "seun-0.23.7", features = ["http-client", "websocket-client"] }
+tendermint-testgen = { git = "https://github.com/composableFi/tendermint-rs", branch = "seun-0.23.7" } # Needed for generating (synthetic) light blocks.
+# Beefy Light Client testing dependencies
+sp-io = { git = "https://github.com/paritytech/substrate", branch = "polkadot-v0.9.22"}
```
Comment (adizere) [inline modules/Cargo.toml:102]: Same for these dev depedendencies.

File: modules/src/core/ics03_connection/handler/conn_open_confirm.rs [hunk 2]
```diff
@@ -1,18 +1,19 @@
...
 use crate::core::ics03_connection::msgs::conn_open_confirm::MsgConnectionOpenConfirm;
+use crate::core::ics26_routing::context::ReaderContext;
 use crate::events::IbcEvent;
 use crate::handler::{HandlerOutput, HandlerResult};
 use crate::prelude::*;
 
-pub(crate) fn process(
-    ctx: &dyn ConnectionReader,
+pub(crate) fn process<HostFunctions: HostFunctionsProvider>(
+    ctx: &dyn ReaderContext,
```
Comment (adizere) [inline modules/src/core/ics03_connection/handler/conn_open_confirm.rs:16]: Why is it important to use the `ReaderContext` here? It seems like `ConnectionReader` was good; it's unclear which changes in the `process` handler warranted switching to `ReaderContext`.

**Comment by Wizdave97** [comment]
> We'll merge #2280 first, so this PR might need very minimal adjustments after that PR lands in master. In the meantime, is it possible to look into the failing CI @Wizdave97? We'll look at failed CI jobs as well and see what's up!

Will do, they are mostly because we've not updated the relayer code, to match some of the changes we made.

File: modules/src/core/ics03_connection/handler/conn_open_confirm.rs (hunk 2 above)
Comment (Wizdave97) [reply modules/src/core/ics03_connection/handler/conn_open_confirm.rs:16]: We didn't like the duplicate methods in different context traits, for example `cient_consensus_state` appearing in `ClientReader`, `ChannelReader` and `ConnectionReader`, so we removed them, and created the `ReaderContext` super trait to capture all that is needed to read anything from the store.

File: modules/Cargo.toml [hunk 3]
```diff
@@ -44,21 +57,36 @@ flex-error = { version = "0.4.4", default-features = false }
...
 
 [dependencies.tendermint-proto]
-version = "=0.23.7"
+git = "https://github.com/composableFi/tendermint-rs"
+branch = "seun-0.23.7"
 default-features = false
 
 [dependencies.tendermint-light-client-verifier]
-version = "=0.23.7"
+git = "https://github.com/composableFi/tendermint-rs"
```
[the remark this answers is not in the document]
Comment (Wizdave97) [reply modules/Cargo.toml:83]: No, That's all that's needed.
I'll open PRs for our changes  on `tendermint-rs` and `ics23` repos.

File: modules/src/core/ics03_connection/handler/conn_open_confirm.rs (hunk 2 above)
Comment (Wizdave97) [reply modules/src/core/ics03_connection/handler/conn_open_confirm.rs:16]: Plus our beefy light client needed some methods that were not available on the `ConnectionReader`

**Comment by mzabaluev** [comment]
Substrate support adds a lot of dependencies to our workspace. Furthermore, many of these must be managed as git dependencies, which would prevent the dependent crates from being published on crates.io.

As stated earlier on Slack, a better path forward for us would be to open up the relayer architecture, so that the Substrate backend could be developed in a separate workspace.

**Comment by adizere** [comment]
Note: blocked on issue #2335. The present PR will need rebasing once that issue is fixed; this will allow to limit the need for substrate dependencies to only the `substrate-light-client` crate, thus keeping the `ibc` crate more minimal in terms of deps.

**Pull request closed without merging**